File: src/main/java/com/kairlec/local/utils/UserUtils.java
package com.kairlec.local.utils;

import com.kairlec.exception.ServiceError;
import com.kairlec.pojo.Captcha;
import com.kairlec.pojo.User;
import com.kairlec.utils.CaptchaMaker;
import com.kairlec.utils.LocalConfig;
import com.kairlec.utils.Network;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;
import java.time.LocalDateTime;

public abstract class UserUtils {
    private static Logger logger = LogManager.getLogger(UserUtils.class);

    private UserUtils() {
        throw new AssertionError("工具类不允许实例化");
    }

    public static void updateLoginInfo(User user, HttpServletRequest request) {
        user.setLastLoginTime(LocalDateTime.now());
        user.setLastSessionId(request.getSession().getId());
    }

    public static void updateLoginInfo(User user, HttpSession session) {
        user.setLastLoginTime(LocalDateTime.now());
        user.setLastSessionId(session.getId());
    }

    public static ServiceError checkStatus(HttpSession session) {
        User user = (User) session.getAttribute("user");
        if (user == null) {
            logger.info("新打开session或未找到已有用户");
            logger.info("无法处理的未登录请求");
            return ServiceError.NOT_LOGGED_IN;
        } else {
            logger.info("在Session取到了'" + user.getUsername() + "'的用户名");
            User targetUser = LocalConfig.getUserService().getUser(user.getUsername());
            if (!targetUser.getLastSessionId().equals(session.getId())) {
                session.invalidate();
                return ServiceError.EXPIRED_LOGIN;
            }
            UserUtils.updateLoginInfo(targetUser, session);
            session.setAttribute("user", targetUser);
            session.setMaxInactiveInterval(60 * 60);
            return ServiceError.NO_ERROR;
        }
    }

    public static ServiceError checkLogin(HttpServletRequest request) {
        HttpSession session = request.getSession();
        if (session.getAttribute("user") != null) {
            return ServiceError.HAD_LOGGED_IN;//已经登录过
        }
        String username = request.getParameter("username");
        String password = request.getParameter("password");
        String IP = Network.getIpAddress(request);
        if (username == null) {
            return ServiceError.NULL_USERNAME;//空的用户名
        }
        if (password == null) {
            return ServiceError.NULL_PASSWORD;//空的密码
        }
        User user = LocalConfig.getUserService().getUser(username);
        if (user == null) {
            return ServiceError.USERNAME_NOT_EXISTS;//错误的用户名
        }
        Captcha captcha = (Captcha) session.getAttribute("captcha");
        if (captcha != null) {
            String captchaString = request.getParameter("captcha");
            if (captchaString == null || !captcha.check(captchaString)) {
                return ServiceError.WRONG_CAPTCHA;//错误的验证码
            }
            session.removeAttribute("captcha");
        } else {
            if (IP != null && user.getIP() != null && !IP.equals(user.getIP())) {
                session.setAttribute("captcha", CaptchaMaker.getCaptcha(LocalConfig.getConfigBean().getCaptchacount()));
                return ServiceError.NEED_VERIFY;//不受信任的IP,需要验证
            }
        }
        password = password.replace(' ', '+');
        try {
            password = PasswordCoderUtils.fromRequest(password);
        } catch (Exception e) {
            return ServiceError.UNKNOWN_PASSWORD;//未知的密码串
        }
        logger.info("解密密码得:" + password);
        logger.info("数据库的密码:" + user.getPassword());
        if (!user.equalsPassword(password)) {
            session.setAttribute("captcha", CaptchaMaker.getCaptcha(LocalConfig.getConfigBean().getCaptchacount()));
            return ServiceError.WRONG_PASSWORD;//错误的密码
        }
        return ServiceError.NO_ERROR;
    }

}

File: src/main/java/com/kairlec/local/utils/UserUtils.kt
package com.kairlec.local.utils

import com.kairlec.config.editable.EditableConfig
import com.kairlec.config.startup.StartupConfigFactory
import com.kairlec.dao.ConfigDao
import com.kairlec.pojo.Captcha
import com.kairlec.exception.ServiceErrorEnum
import com.kairlec.local.utils.PasswordCoderUtils.fromRequest
import com.kairlec.pojo.User
import com.kairlec.utils.*
import org.apache.logging.log4j.LogManager
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpSession

object UserUtils {
    private val logger = LogManager.getLogger(UserUtils::class.java)

    private val startupConfig
        get() = StartupConfigFactory.Instance

    fun authHttpSession(session: HttpSession): ServiceErrorEnum {
        (session.getAttribute("user") as? User)?.let { user ->
            logger.info("""在Session取到了"${user.username}"的用户名""")
            return when {
                //ID不一样,在其他地方登录,挤下线
                user.lastSessionId != session.id -> {
                    session.invalidate()
                    ServiceErrorEnum.EXPIRED_LOGIN
                }
                else -> {
                    ServiceErrorEnum.NO_ERROR
                }
            }
        }
        return ServiceErrorEnum.NOT_LOGGED_IN
    }

    fun authHttpServletRequest(request: HttpServletRequest, blackAPIList: Array<String>): ServiceErrorEnum {
        val requestUrl = URLDecoder.decode(request.requestURI, StandardCharsets.UTF_8)
        if (requestUrl in blackAPIList) {
            return ServiceErrorEnum.NO_ERROR.data("BlackList")
        }
        if (!request.method.equals("POST", ignoreCase = true)) {
            return ServiceErrorEnum.UNKNOWN_REQUEST
        }
        val session = request.getSession(true)
        return authHttpSession(session)
    }

    fun authLogin(request: HttpServletRequest): User {
        val session = request.session
        if (session.getAttribute("user") != null) {
            ServiceErrorEnum.HAD_LOGGED_IN.throwout() //已经登录过
        }
        val username = request.getParameter("username")
        var password = request.getParameter("password")
        val ip = request.IP
        if (username == null) {
            ServiceErrorEnum.NULL_USERNAME.throwout() //空的用户名
        }
        if (password == null) {
            ServiceErrorEnum.NULL_PASSWORD.throwout() //空的密码
        }
        val user = EditableConfig.getUser(username)
                ?: ServiceErrorEnum.USERNAME_NOT_EXISTS.throwout() //错误的用户名
        val captcha = session.getAttribute("captcha")
        if (captcha is Captcha) {
            val captchaString = request.getParameter("captcha")
            if (captchaString == null || !captcha.check(captchaString)) {
                ServiceErrorEnum.WRONG_CAPTCHA.throwout() //错误的验证码
            }
            session.removeAttribute("captcha")
        } else {
            if (ip != user.ip) {
                session.setAttribute("captcha", Captcha.getInstant(startupConfig.captchaCount))
                ServiceErrorEnum.NEED_VERIFY.throwout() //不受信任的IP,需要验证
            }
        }
        password = password.replace(' ', '+')
        password = fromRequest(password)
        if (!user.equalsPassword(password)) {
            session.setAttribute("captcha", Captcha.getInstant(startupConfig.captchaCount))
            ServiceErrorEnum.WRONG_PASSWORD.throwout() //错误的密码
        }
        user.ip = ip
        user.lastSessionId = session.id
        EditableConfig.save()
        return user
    }

    fun createCookie(user: User, request: HttpServletRequest, expiredTime: Long): String {
        return RSACoder.encryptByPublicKeyToString("${VerifyUtils.getStringVerifyAsHex(user.username, VerifyAlgorithmEnum.MD5)};${request.IP};${System.currentTimeMillis() + expiredTime};${user.salt}", ConfigDao.Instance.publicKey)
    }

    fun authCookie(cookie: String, request: HttpServletRequest): User? {
        try {
            val decodedCookie = RSACoder.decryptByPrivateKeyToString(cookie, ConfigDao.Instance.privateKey)
            val cookieArr = decodedCookie.split(';')
            if (cookieArr.size != 4) {
                return null
            }
            val user = EditableConfig.config.adminUserArray.find { VerifyUtils.verifyString(it.username, VerifyAlgorithmEnum.MD5, cookieArr[0]) }
                    ?: return null
            if (request.IP != cookieArr[1]) {
                return null
            }
            val expiredTime = cookieArr[2].toLongOrNull() ?: return null
            if (expiredTime < System.currentTimeMillis()) {
                return null
            }
            if (cookieArr[3] != user.salt) {
                return null
            }
            return user
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        }
    }
}
